### PR TITLE
[IOBP-524] Allow user to switch apps while authenticating on Android

### PR DESCRIPTION
**Pull request by mastro993**
## Short description
This PR fixes an issue on Android devices where users were unable to switch apps during the authentication session. With this fix the browser is able to remain active while the user navigates to other apps.

The solution takes inspiration from a similar issue encountered by AppAuth for Android library ([here the PR](https://github.com/openid/AppAuth-Android/pull/109)). It uses an intermediary Activity, `AuthenticationManagerActivity`, to allow the Chrome Custom Tab to be started without the `NEW_ACTIVITY_NO_HISTORY` flag, and the `RedirectUriReceiverActivity`, which allows to remove the Browser activity from the recents.

Untitled-2024-01-31-1157

<details><summary>Mermaid Flow</summary>
<p>

```mermaid
sequenceDiagram
    participant ReactAppMainActivity
    participant openAuthenticationSession as openAuthenticationSessio(io-react-native-login-utils)
    participant AuthorizationService as AuthorizationService(io-react-native-login-utils)
    participant BrowserHandler as BrowserHandler(io-react-native-login-utils)
    participant CustomTabsServiceConnection as CustomTabsServiceConnection(io-react-native-login-utils)
    participant AuthManagerActivity as AuthManagerActivity(io-react-native-login-utils)
    participant ChromeCustomTabs
    participant RedirectUriActivity as RedirectUriActivity(io-react-native-login-utils)

    ReactAppMainActivity->>openAuthenticationSession: Calls with auth URL
    openAuthenticationSession->>AuthorizationService: Instantiates
    AuthorizationService->>BrowserHandler: Instantiates
    BrowserHandler->>CustomTabsServiceConnection: Instantiates
    BrowserHandler->>ReactAppMainActivity: Do bind service
    ReactAppMainActivity->>CustomTabsServiceConnection: Bind    
    AuthorizationService->>AuthManagerActivity: performAuthRequest with authIntent and start    
    AuthorizationService->>AuthorizationService: Disposing...
    AuthorizationService->>BrowserHandler: Unbind        
    BrowserHandler->>ReactAppMainActivity: Do unbind service
    ReactAppMainActivity->>CustomTabsServiceConnection: Unbind
    AuthorizationService->>AuthorizationService: Disposed
    AuthManagerActivity->>ChromeCustomTabs: Starts authIntent
    Note right of AuthManagerActivity: Saves state of authIntent and tracks if auth request has previously started.
    ChromeCustomTabs->>RedirectUriActivity: The auth flow ends with an app link that trigger
    RedirectUriActivity->>AuthManagerActivity: Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP
    RedirectUriActivity->>RedirectUriActivity: Kills itself    
    AuthManagerActivity->>AuthManagerActivity: Handles response
    Note over AuthManagerActivity: Restore state of authIntent and if an auth request has previously started.
    AuthManagerActivity->>ReactAppMainActivity: resolve/reject promise
    AuthManagerActivity->>AuthManagerActivity: kills itself thus removing ChromeCustomTabs from history

    Note over ReactAppMainActivity, RedirectUriActivity: Flow initiates in ReactAppMainActivity and transitions through library components to manage authentication via ChromeCustomTabs. RedirectUriActivity handles the final redirect URI to adjust the navigation stack

```

</p>
</details> 

## List of changes proposed in this pull request
- Added `AuthorizationManagerActivity` which is responsible to prepare and launch the auth Intent (`CustomTabsIntent`)
- Added `RedirectUriReceiverActivity` which handles the result from the `CustomTabsIntent` via deeplink
- Improved error management

## How to test
On **Android**, within the example app, start the [io-dev-api-server](https://github.com/pagopa/io-dev-api-server) and try to login and/or perform a payment request. You should be able to navigate to other apps while the Chrome custom tab is open and  you should be able to complete the authentication/authorization when returning back to the example app.

## Preview

<video src="https://github.com/pagopa/io-react-native-login-utils/assets/6160324/c435df3c-4939-4e6b-b1bb-212b0b853f4f" />

### CODE REVIEW

File: android/src/main/java/com/iologinutils/AuthorizationManagerActivity.kt [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  private var authIntent: Intent? = null
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+
+    intent.extras?.let {
+      extractState(it)
+    } ?: savedInstanceState?.let {
+      extractState(it)
+    } ?: throw IllegalStateException("No state to extract")
```
Comment (shadowsheep1) [inline android/src/main/java/com/iologinutils/AuthorizationManagerActivity.kt]: We need to find a better (more gentle) way to handle this situation. Maybe rejecting with a specific error. Better than crashing the whole app :). What do you think?

File: android/src/main/java/com/iologinutils/AuthorizationManagerActivity.kt [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  private fun extractState(state: Bundle) {
+    state.apply {
+      authIntent = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
+        state.getParcelable(KEY_AUTH_INTENT, Intent::class.java)
+      } else {
+        @Suppress("DEPRECATION")
+        state.getParcelable(KEY_AUTH_INTENT)
+      }
```
Comment (shadowsheep1) [inline android/src/main/java/com/iologinutils/AuthorizationManagerActivity.kt:151]: I handled the deprecation. Do you have a device running Android 13 (API level 33) to test the new API?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]
More to come... ;)

File: android/src/main/java/com/iologinutils/AuthorizationManagerActivity.kt (hunk 1 above)
Comment (mastro993) [reply android/src/main/java/com/iologinutils/AuthorizationManagerActivity.kt]: You are right, we should reject with a platform specific error. 
We could extend the `AndroidError` type with an error for this case and then reject the `Promise` with it.

Comment (mastro993) [reply android/src/main/java/com/iologinutils/AuthorizationManagerActivity.kt]: I've added an `IllegalStateException` rejection error in https://github.com/pagopa/io-react-native-login-utils/pull/11/commits/86956be4cd96830eda1f4d4d191137e182afb02a, let me know what do you think!

Comment (shadowsheep1) [reply android/src/main/java/com/iologinutils/AuthorizationManagerActivity.kt]: Okay, perfect!

File: src/index.tsx [hunk 3]
```diff
@@ -19,42 +19,36 @@ const IoLoginUtils = NativeModules.IoLoginUtils
...
-  userInfo: {
-    Error: Error;
-    URL: string | undefined;
-    StatusCode: number | undefined;
-    Parameter: Array<string> | undefined;
+  userInfo?: {
+    error: Error;
+    url: string | undefined;
+    statusCode: number | undefined;
+    parameter: Array<string> | undefined;
```
Comment (shadowsheep1) [inline src/index.tsx:51]: These are breaking changes, if we made those modifications we need to update the major version of the lib and describe the changes in the changelog of the new release. What do you think?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

File: src/index.tsx (hunk 3 above)
Comment (mastro993) [reply src/index.tsx:51]: Absolutely.
This change aligns the property names of `userInfo` (PascalCase) to the code style of the RN part (camelCase). Also, some errors may not have user info, so `userInfo` becomes nullable.

**Review by shadowsheep1 — APPROVED** [review]
LGTM 🚀. 

🚨 Just don't forget to bump the major version when you make the new release and to add in the release description a section explaining that there are breaking changes and what they are.

**Pull request merged**